### PR TITLE
Fixed #4

### LINKED ISSUE #4: Exceeding the limit set by the user
``` python
In [1]: from ghscrap.scrapper import search_location
In [2]: spam = search_location("Kharagpur", max_users=15)
In [3]: len(spam)
Out[3]: 20
```

Currently, the number of users returned by the function `search_location` exceeds the limit set by `max_users`.  We need to implement this restriction.

### PR DESCRIPTION
**Pull request by zorroblue**
Will research a little more on #2 .
This commit fixes #4

### CODE REVIEW

**Comment by kshitij10496** [comment]
@zorroblue
1. I would prefer the check to be placed in `search_location` (or its helper). Can you explain the reason behind your approach ?
2. This commit attempts to fix 2 issues simultaneously. Please separate the fixes.
   Also, I would appreciate if you collaborate with @Arafatk over fixing #2.

**Comment by zorroblue** [comment]
1. I didn't get you here. My approach was pretty simple: in the for loop where you get users from each page, I did an update to reduce max_users after every time I finish off a page. So this sets the limit for the new page.
   An alternative was introducing a global variable. But I felt it might lead to more changes in the code and make the code lose readability.
2. I should have separated the commits, my mistake :)

**Comment by arafatkatze** [comment]
This seems reasonable when I check it with 

```
for index in range(10,30):
  users = search_location("Kharagpur", max_users = index)
  print len(users)
```

Another simple way would be to use a [deque](https://docs.python.org/2/library/collections.html#deque-objects) and set it maxlen parameter(and break when deque size reaches maxlen). Although I am not sure if its 'better'.
Also instead of `int(ceil(num))+1` we could also do `int(num)+2` and avoid ceil altogether. Just a thought.

**Comment by kshitij10496** [comment]
> Also instead of int(ceil(num))+1 we could also do int(num)+2 and avoid ceil altogether. Just a thought.

This is a PR for separate issue.
Let's discuss it at a more suitable place. :smile:

**Comment by zorroblue** [comment]
@Arafatk @kshitij10496 Am pushing my thoughts for ^^ and ^ over to #2 discussion

**Comment by kshitij10496** [comment]
Here are a few comments on how to modify this PR:
1. Undo your commit but keep the modified changes
2. Separate the changes into 2 commits.
3. Undo the commit related to #2 (As @Arafatk is working on it)
4. I would prefer to modify the behaviour of `search_location` function for handling this issue.
   `search_page_users` is true to its purpose of returning "list of all the users on a search result page".

If you need help with Git, Stack Overflow is your friend !
If you still can't figure out the procedure, ask on the [Gitter Channel](https://gitter.im/kshitij10496/gh-scrapper).

**Comment by kshitij10496** [comment]
Ping @zorroblue 
What's the update on this ?

**Comment by zorroblue** [comment]
I am really sorry @kshitij10496 I came home now only. Will work and update you by tomorrow :)

**Comment by zorroblue** [comment]
Hey, I saw #1 fixed. Do I git pull the new source and work on it instead?

**Comment by kshitij10496** [comment]
@zorroblue Thats absolutely fine ! :smile:

**Comment by kshitij10496** [comment]
Do this on your `master`
1. Check your `remote` first : `git remote -v`
2. Add this repo as an `upstream`: `git remote add upstream https://github.com/kshitij10496/gh-scrapper.git`
3. Fetch the changes and merge: 
`git fetch upstream  
git merge upstream/master`
(I don't suggest to use `git pull`)

Checkout this branch : `git checkout limitexceed`
1. Rebase this branch onto master: `git rebase master`
2. If there are **Merge commits**, then fix them.
Google is your friend. :wink: 
3. Update your branch to your suiting.
4. Push the changes `git push origin limitexceed`.

If you face any other problem, comment here or we can discuss it on the Gitter channel.

**Comment by zorroblue** [comment]
Why don't you suggest git pull ? I thought it was a one-step solution :) 
Is there something wrong behind it? I have only used it for small projects, never had problems.

**Comment by zorroblue** [comment]
Hey and Btw I am gonna save your comment in my reference..
You have a great future in StackOverflow,my friend! :D

**Comment by kshitij10496** [comment]
I have 2 primary reasons for advising against the use of `git pull`:
1. The problem with `git pull` is that it has all kinds of helpful "magic" (like you suggested yourself).
   It means that you don’t really get to learn about the branching in git. Mostly things 'Just Work', but when they don’t it’s often difficult to work out why.
2. The other problem is that by both fetching and merging in one command, your working directory is updated without giving you a chance to examine the changes you’ve just brought into your repository.
   At times, this can sometimes lead to a difficult situation.

My first ever PR was messed up due to `git pull`. So there's that as well ! :stuck_out_tongue_winking_eye:

**Comment by zorroblue** [comment]
@kshitij10496  With regard to the function naming of search_page_users, what you say makes perfect sense. But in that case , why do you pass max_users to the search_page_users function?

**Comment by kshitij10496** [comment]
> But in that case , why do you pass max_users to the search_page_users function?

I don't think that I pass the `max_users` argument to `search_page_users`.
I am -1 on passing this arg to this functionality.

**Comment by zorroblue** [comment]
Yeah you are right.
I will look into refactoring it in that way. Sounds better :+1:

**Comment by kshitij10496** [comment]
Cool ! :+1:

**Comment by zorroblue** [comment]
If _search_page_users has to parse the entire page, how do you think we should target the 'partial scraping' of data?

**Comment by zorroblue** [comment]
I got a way now :) Ignore ^
My suggestion:
Update max_users as I did before..when it falls below 10, trim the users list got from _search_page_users in the _search_location

**Comment by kshitij10496** [comment]
> If _search_page_users has to parse the entire page, how do you think we should target the 'partial scraping' of data?

`search_page_users` should scrape data from the entire page. 
`search_location` need not.

**Comment by zorroblue** [comment]
@kshitij10496 Got it!

**Comment by zorroblue** [comment]
@kshitij10496  I did the necessary changes. :)

File: ghscrap/scrapper.py [hunk 1]
```diff
@@ -113,7 +100,13 @@ def _search_location(soup, max_users, params):
 
     users = []
     for i in range(1, number_of_pages):
-        users.extend(search_page_users(i, params=params))
+        parsed_users=search_page_users(i, params=params)
```
Comment (kshitij10496) [inline ghscrap/scrapper.py]: Its a valid solution to the problem.
However, I think this complicates the process too much.

Adding a single check to length of list `users` should be enough !

For example:
If the user wants 15 names, we can scrape 20 and return the first 15.
Thats fine by me.

File: ghscrap/scrapper.py [hunk 2]
```diff
@@ -113,7 +100,13 @@ def _search_location(soup, max_users, params):
...
+        if max_users<10:
+            users.extend(parsed_users[:max_users])
+            break
+        else:
+            users.extend(parsed_users)
+            max_users=max_users-10
```
Comment (kshitij10496) [inline ghscrap/scrapper.py]: It's not a good design technique to change the value of a parameter internally.

File: ghscrap/scrapper.py [hunk 3]
```diff
@@ -1,3 +1,4 @@
+
```
Comment (kshitij10496) [inline ghscrap/scrapper.py]: Remove this line

File: ghscrap/scrapper.py [hunk 4]
```diff
@@ -16,25 +17,19 @@ def kitchen(params):
 
 def user_search_filter(filterkey):
     """Filter for user search pages.
-
```
Comment (kshitij10496) [inline ghscrap/scrapper.py:19]: Add all these deleted blank lines.
They are according to coding style guide [PEP8](https://www.python.org/dev/peps/pep-0008/)

File: ghscrap/scrapper.py [hunk 5]
```diff
@@ -153,4 +146,4 @@ def main():
 
 
 if __name__ == '__main__':
-    main()
+    main()
```
Comment (kshitij10496) [inline ghscrap/scrapper.py]: There should be an EOF at the end of every module.

**Comment by zorroblue** [comment]
Done!

**Comment by zorroblue** [comment]
I complicated things a lot. I realised it only after seeing your simple solution.

File: ghscrap/scrapper.py [hunk 6]
```diff
@@ -16,10 +16,10 @@ def kitchen(params):
 
 def user_search_filter(filterkey):
     """Filter for user search pages.
-
+
```
Comment (kshitij10496) [inline ghscrap/scrapper.py]: Remove the indented spaces in all the lines.

**Comment by kshitij10496** [comment]
> I complicated things a lot. I realised it only after seeing your simple solution.

No problem !

**Comment by zorroblue** [comment]
I will correct it asap. I don't remember doing anything with the line spaces or indentation. What could have been the reason behind this??

**Comment by kshitij10496** [comment]
I would suggest you to read about PEP8 coding style guide.
Almost all the Python projects I have come across follow this guide.

It would be better if you can install some plugin for checking the coding style automatically for you.

**Comment by kshitij10496** [comment]
> What could have been the reason?

 Are you using Sublime Text as your text editor ?

**Comment by zorroblue** [comment]
Yes :O

**Comment by kshitij10496** [comment]
In Sublime Text, the default behaviour when you press Enter in a python module is to auto indent it to 4 spaces(or whatever number you have set. Check your status bar.)

**Comment by kshitij10496** [comment]
Good work !
This is in.
Thanks for your contribution.

**Comment by zorroblue** [comment]
@kshitij10496 What's next?

**Comment by kshitij10496** [comment]
@zorroblue Today I am focussing on developing tests for the functions we have implemented as of now.

**Comment by zorroblue** [comment]
Where can I step in? Can I work on feature 2? Wait, let's chat on Gitter instead!

**Pull request merged**